services/paymaster/src/rpc.rs

```diff
@@ -153,41 +156,109 @@ impl ChainIndex {
 /// Sends a transaction and waits for confirmation via WebSocket subscription.
 #[tracing::instrument(
     skip_all,
-    fields(tx_hash = %transaction.signatures[0])
+    fields(
+        tx_hash = %transaction.signatures[0],
+        result
+    )
 )]
 pub async fn send_and_confirm_transaction(
     rpc: &RpcClient,
-    pubsub: &PubsubClient,
+    pubsub: &PubsubClientWithReconnect,
     transaction: &VersionedTransaction,
     config: RpcSendTransactionConfig,
 ) -> Result<ConfirmationResultInternal, ErrorResponse> {
     let signature = match rpc.send_transaction_with_config(transaction, config).await {
-        Ok(sig) => sig,
+        Ok(sig) => {
+            tracing::Span::current().record("result", "sent");
+            sig
+        }
         Err(err) => {
             if let Some(error) = err.get_transaction_error() {
+                tracing::Span::current().record("result", "preflight_failure");
+                tracing::warn!(
+                    "Transaction {} failed preflight: {error}",
+                    transaction.signatures[0]
+                );
                 return Ok(ConfirmationResultInternal::UnconfirmedPreflightFailure {
                     signature: transaction.signatures[0],
                     error,
                 });
             }
+            tracing::Span::current().record("result", "send_failed");
+            tracing::error!(
+                "Failed to send transaction {}: {err}",
+                transaction.signatures[0]
+            );
             return Err(to_error_response(err));
         }
     };
 
-    confirm_transaction(pubsub, signature, Some(rpc.commitment())).await
+    confirm_transaction_with_reconnect(pubsub, signature, Some(rpc.commitment())).await
 }
 
 pub const CONFIRMATION_TIMEOUT: Duration = Duration::from_secs(60);
 
 #[tracing::instrument(
     skip_all,
-    fields(tx_hash = %signature)
+    name = "confirm_transaction",
+    fields(
+        tx_hash = %signature,
+        result
+    )
 )]
+async fn confirm_transaction_with_reconnect(
+    pubsub: &PubsubClientWithReconnect,
+    signature: Signature,
+    commitment: Option<CommitmentConfig>,
+) -> Result<ConfirmationResultInternal, ErrorResponse> {
+    let result = {
+        let pubsub_guard = pubsub.client.load();
+        confirm_transaction(&pubsub_guard, signature, commitment).await
+    };
+
+    match result {
+        Ok(confirmation) => {
+            tracing::Span::current().record("result", confirmation.status_string().as_str());
+            Ok(confirmation)
+        }
+        Err(e) if is_subscription_error(e.0, &e.1) => {
+            tracing::warn!("WebSocket subscription failed, attempting reconnection and retry");
+            pubsub.reconnect_pubsub().await?;
+            let new_guard = pubsub.client.load();
+            let retry_result = confirm_transaction(&new_guard, signature, commitment)
+                .await
+                .map_err(|e| e.into());
+
+            match retry_result {
```

Comment:
this nested retry doesn't feel great. It leads to the duplication of:
- tracing::Span::current().record("result", confirmation.status_string().as_str());
- tracing::Span::current().record("result", "unconfirmed");